bzlib.rs

```diff
@@ -591,44 +586,38 @@ unsafe fn BZ2_bzCompressInitHelp(
     let n = 100000 * blockSize100k;
 
     let arr1_len = n as usize;
-    let arr1_alloc = bzalloc_array(*bzalloc, (*strm).opaque, arr1_len);
+    let arr1 = Arr1::alloc(bzalloc, (*strm).opaque, arr1_len);
 
     let arr2_len = n as usize + (2 + 12 + 18 + 2);
-    let arr2_alloc = bzalloc_array(*bzalloc, (*strm).opaque, arr2_len);
+    let arr2 = Arr2::alloc(bzalloc, (*strm).opaque, arr2_len);
 
-    let ftab_alloc = bzalloc_array(*bzalloc, (*strm).opaque, FTAB_LEN);
+    let ftab = Ftab::alloc(bzalloc, (*strm).opaque);
 
-    let (Some(arr1_alloc), Some(arr2_alloc), Some(ftab_alloc)) =
-        (arr1_alloc, arr2_alloc, ftab_alloc)
-    else {
-        if let Some(ptr) = arr1_alloc {
-            (bzfree)((*strm).opaque, ptr as *mut c_void);
+    match (arr1, arr2, ftab) {
+        (Some(arr1), Some(arr2), Some(ftab)) => {
+            (*s).arr1 = arr1;
+            (*s).arr2 = arr2;
+            (*s).ftab = ftab;
+        }
+        (arr1, arr2, ftab) => {
+            if let Some(mut arr1) = arr1 {
+                arr1.dealloc(bzfree, (*strm).opaque);
         }
 
-        if let Some(ptr) = arr2_alloc {
-            (bzfree)((*strm).opaque, ptr as *mut c_void);
+            if let Some(mut arr2) = arr2 {
+                arr2.dealloc(bzfree, (*strm).opaque);
         }
 
-        if let Some(ptr) = ftab_alloc {
-            (bzfree)((*strm).opaque, ptr as *mut c_void);
+            if let Some(mut ftab) = ftab {
+                ftab.dealloc(bzfree, (*strm).opaque);
         }
 
-        if !s.is_null() {
```

Comment:
We are already dereferencing this pointer before this point anyway, so `s` has to be a non-NULL pointer to not have UB. No need to check again.